File: profiling/runner_reap.py
import argparse
import csv
import logging
import time

import utils.profile as memory_profiler
import utils.constants as constants

def main(language: str, experiment: str, loop: int) -> None:
    with open(f"{constants.LOG_PATH}/reap_{language}_{experiment}_memory.csv", mode='w') as file:
        writer = csv.writer(file)
        writer.writerow(["Memory"])
        
        for i in range(loop):
            logging.info(f"Loop: {i+1}/{loop}")
            memory_page = []
            profiler = memory_profiler.MemoryProfiler(language, experiment, None, None)
            profiler.gracefuly_stop_fc_vm()
            profiler.start_new_vm()
            
            profiler.run_server()
            profiler.warmup_server(10)
            profiler.take_snapshot()
            profiler.load_snapshot()
            
            while True:
                return_code = profiler.invoke_server()
                if return_code == 0:
                    break
                logging.info("Retrying ...")
            logging.info("response returned")
            memory_page.append(profiler.get_current_memory_page())
            logging.info(f"Req memory page: {memory_page[0]}")
            
            profiler.gracefuly_stop_fc_vm()
            
            writer.writerow(memory_page)


if __name__ == '__main__':
    logging.basicConfig(level=logging.INFO)
    parser = argparse.ArgumentParser(description='Profile the code')
    parser.add_argument('--language', '-l', choices= ["python", "cpp", "node"], help='Program language')
    parser.add_argument('--expr', '-e', type=str, help='Experiment name')
    parser.add_argument('--loop', '-n', type=int, help='Number of experiment loop')
    args = parser.parse_args()
    
    main(args.language, args.expr, args.loop)

File: go_profiling/runner_reap.go
package main

import (
	"encoding/csv"
	"flag"
	"os"
	"strconv"
	"time"

	log "github.com/sirupsen/logrus"

	"github.com/JooyoungPark73/fc_benchmark/go_profiling/utils"
)

func RunReap() {
	// Command-line arguments
	language := flag.String("language", "", "Program language")
	experiment := flag.String("expr", "", "Experiment name")
	loop := flag.Int("loop", 1, "Number of experiment loop")
	flag.Parse()

	// CSV file for results
	filePath := utils.Log_Path + "/reap_" + *language + "_" + *experiment + "_memory.csv"
	file, err := os.Create(filePath)
	if err != nil {
		log.Fatalf("Failed to create CSV file: %v", err)
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	// Write header
	writer.Write([]string{"Memory"})

	// Loop through experiments
	for i := 0; i < *loop; i++ {
		log.Infof("Loop: %d/%d\n", i+1, *loop)
		memoryPage := []string{}

		profiler, _ := utils.NewMemoryProfiler(*language, *experiment, "", "")
		profiler.GracefullyStopFCVM()
		profiler.StartNewVM()

		profiler.RunServer()
		profiler.WarmupServer(10)
		profiler.TakeSnapshot()
		profiler.LoadSnapshot()

		// Retry until server returns successfully
		for {
			output, err := profiler.InvokeServer()
			if err == nil {
				log.Infof("Server response: %s\n", output)
				break // Exit the loop on success
			}

			log.Infof("Error invoking server: %v\n", err)
			log.Infof("Retrying ...")
			time.Sleep(5 * time.Second) // Add a small delay before retrying
		}
		log.Infof("Response returned")
		memoryPage = append(memoryPage, strconv.Itoa(profiler.GetCurrentMemoryPage()))
		log.Infof("Req memory page: %s\n", memoryPage[0])

		profiler.GracefullyStopFCVM()

		// Write results to CSV
		writer.Write(memoryPage)
	}
}
